### PR TITLE
[WIP] Support LLL compilation in StandardCompiler

### PR DESCRIPTION
**Pull request by axic**
TODO:
- [ ] tests
- [ ] same format for input as Solidity
- [x] same format for ouput as Solidity
- [x] change LLL `(include ..)` to be restricted (#3008)

### CODE REVIEW

File: libsolidity/interface/StandardCompiler.cpp [hunk 1]
```diff
@@ -160,6 +160,38 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
...
+		return formatFatalError("JSONError", "Source missing.");
+
+	Json::Value const& settings = _input.get("settings", Json::Value());
+	Json::Value optimizerSettings = settings.get("optimizer", Json::Value());
+	bool optimize = optimizerSettings.get("enabled", Json::Value(false)).asBool();
+
+	vector<string> errorList;
+	Json::Value output;
+	output["contracts"] = Json::objectValue;
+	output["contracts"][""] = Json::objectValue;
```
Comment (axic) [inline libsolidity/interface/StandardCompiler.cpp:168]: Question is to do we go for
```
{
  "contracts": {
    "": {
      "bytecode": ""
    }
  }
}
```

in order to be "compatible" with the usual, or go with:

```
{
  "bytecode": "..."
}
```

**Comment by chriseth** [comment]
Note that lll is currently disabled for emscripten, I think because it uses some more boost libraries that are quite large (e.g. spirit). Let's see how it compiles.

**Comment by zigguratt** [comment]
I just took a look at what spirit does in the LLL compiler (hey, I'm new to this C++ stuff!) and I can't see removing its use without a total rewrite of the parser, at the very least. I'm just hoping that enabling it for emscripten will be practical. The world needs LLL! :)

**Comment by axic** [comment]
Fixed compilation, now we can see if it works with emscripten.

File: libsolidity/interface/StandardCompiler.cpp [hunk 2]
```diff
@@ -160,6 +161,39 @@ Json::Value collectEVMObject(eth::LinkerObject const& _object, string const* _so
 	return output;
 }
 
+Json::Value compileLLL(Json::Value const& _input)
+{
+	if (!_input["source"].isString())
```
Comment (axic) [inline libsolidity/interface/StandardCompiler.cpp:158]: We've agreed on a call some time ago that the input should be the same format as for Solidity. This needs to be changed to reflect that.

**Comment by axic** [comment]
This is now also blocked by LLL's `(include)` which has no restrictions to access files.

File: libsolidity/interface/StandardCompiler.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/interface/StandardCompiler.cpp:168]: Fixed.

**Comment by zigguratt** [comment]
> @axic , couldn't this one be closed (due to the announced deprecation in https://solidity.readthedocs.io/en/v0.5.0/lll.html) ?

That's sad news. I had no idea LLL was up for deprecation and eventual removal from the Solidity repository. A lot of work has gone into [documenting](https://lll-docs.readthedocs.io/en/latest/) and [explaining](http://bit.ly/2F8qBIG) the language. The first Medium article I wrote has over 12,000 views with a 46% read ratio. LLL is arguably still the most efficient language on the EVM, producing the smallest binaries.

**Comment by lazaridiscom** [comment]
@zigguratt , the main issue is:

LLL does not belong within the Solidity repo (see issue #5393).

This is what the "deprecation" is about (not about killing LLL).

I understand that you are a big fan of LLL. A few people should be able to decouple the source and move it over to an own repository.

If LLL is so good as you say, it deserves anyways an own repo.

**Comment by zigguratt** [comment]
@lazaridiscom 

Not only is it as good as I say :) but it's being used in a couple of high-profile projects in their smaller, often-invoked contracts for the very efficiency I mentioned. Also, it's finally possible to add support for LLL in Truffle, so many more people will be exposed to the language as a choice there.

> LLL does not belong within the Solidity repo (see issue #5393).

I completely agree with this and have said that for the last couple of years. It would be wonderful if LLL were broken out into a separate repository. Then the language could have its own release schedule, version, etc. I have minimal C++ experience but I would certainly do what I could to help the process along.

From #5393: @axic said "split out a library by strict refactoring, then use it from multiple projects"

That would be the cleanest route to take. Long ago I suggested Git submodules for the parts of the code base that were shared between languages. Submodules were not viewed positively by the Solidity team at that time. I don't know how they're viewed these days.

**Comment by lazaridiscom** [comment]
> axic said "split out a library by strict refactoring, then use it from multiple projects"

That was actually me, within https://github.com/ethereum/solidity/issues/5393#issuecomment-437898786

> I have minimal C++ experience but I would certainly do what I could to help the process along.

There is a tiny chance that I create a skeleton project at some point (mostly out of curiosity to assess the effort). Then it would be easier for you to attract some more folks to attempt the full migration.

**Comment by axic** [comment]
I was considering what should happen to this last week.  I think my current view is that it is a bad idea finishing this, because then removal of it would be a breaking change.

I don't think anyone has the capacity to finish this and I'd say effort should instead be put into finishing the LLL to Yul compiler as a separate project.

**Pull request closed without merging**